non-singular Laplace BIE solver, improved FFT interpolation for singular integral, and new magnetic field API

unalmis opened the pull request:
![PR Age](https://img.shields.io/github/issues/detail/age/PlasmaControl/DESC/1360?label=Ready%20for%20merge%20since)

All the solvers are thoroughly tested. See the shared paper to explain this PR.

This is the only open source JAX AD compatible BIE solver.

The robustness of the free surface optimization is outside scope of this pull request. See #1894 and #1208 for more details. There are plenty of tests that illustrate correctness, and tuning optimization hyper parameter defaults which is problem dependent anyway should not delay merging this PR.

This PR includes my FFT improvements significantly improve free surface optimization (both old and new) and anything that involves singular integrals.

The quadrature in master has been observed to have unfortunately slow convergence for some test cases. It has been shown the improved quadrature in the linked paper will converge faster especially for the non singular kernel of this system. It can be added in a different PR.

unalmis commented:
I'll push the simpler method we talked about. Regardless it is not expected for the current method to not satisfy the lcfs condition. 

Am I missing something in the implementation? If you are available, but would like to be caught up on the code or math let me know.

unalmis committed:
Increase chunking

unalmis committed:
Rename to bs_chunk_size

unalmis committed:
add kwargs

unalmis committed:
Merge branch 'master' into ku/biot

unalmis committed:
Merge branch 'master' into ku/biot

unalmis committed:
Merge branch 'ku/biot' into dp/laplace

unalmis committed:
Updating tests

unalmis committed:
Merge branch 'master' into ku/biot

unalmis committed:
Merge branch 'ku/biot' into dp/laplace

unalmis committed:
Merge branch 'master' into ku_rc/anistropy

unalmis committed:
Make sure theta is on outboard side in test

unalmis committed:
Add more general test

unalmis committed:
Add flag to docstring

dpanici commented on tests/test_integrals.py:
```diff
@@ -667,38 +696,346 @@ def test_singular_integral_vac_estell(self, interpolator):
+            check=False,
+            chunk_size=chunk_size,
+        )
+        assert np.isfinite(Phi_mn).all()
```
try with finite diff, vmap ID matrix, or build the pinv yourself

unalmis replied:
vmap id matrix is better than building pinv

unalmis replied:
Changing to avoid the Jacobian still had nan. After more debugging, the `nan` was another JAX bug. Once the matrix is large enough, JAX just gives up and fills linear algebra solve with nan. I fixed this by chunking the computation, just like in #1522 .

unalmis replied:
Will push fixed commit soon.

unalmis committed:
Switch to vmap identity

unalmis committed:
Use vmap chunked

unalmis committed:
Workaround JAX bug with nan linear algebra solve

unalmis committed:
Reduce tolerance

unalmis commented on tests/test_integrals.py:
```diff
@@ -667,38 +695,334 @@ def test_singular_integral_vac_estell(self, interpolator):
+        )
+        Bn = B0n + dPhi_dn
+        # FIXME: This fails to converge to zero, only converges to 0.03.
+        np.testing.assert_allclose(Bn, 0, atol=0.03)
```
@f0uriest Can you push the axissymetric bugfix to the anisotropy branch? Then we can merge it to this branch. Maybe that will fix the issue.

Here you can see above that `Phi` is computed correctly, but the result does not converge to zero. This is just applying the shielding theorem.

unalmis replied:
It didn't change anything, but thank you.

unalmis commented on desc/integrals/_laplace.py:
```diff
@@ -0,0 +1,572 @@
+
+
+# TODO: remove below stuff before merge.
+# alternative methods to check for agreement
```
can ignore below

f0uriest committed:
Fix heuristic params for axisymmetry

unalmis committed:
Merge branch 'ku_rc/anistropy' into dp/laplace

unalmis committed:
move docstring to correct test

unalmis committed:
cosmetic change to avoid confusion

unalmis commented on tests/test_integrals.py:
```diff
@@ -667,38 +692,334 @@ def test_singular_integral_vac_estell(self, interpolator):
+            warn_fft=False,
+        )
+        Bn = B0n + dPhi_dn
+        # FIXME: This fails to converge to zero, only converges to 0.03.
```
Analytically, we know that $\nabla \Phi \cdot n = - sin(\theta)$. (You can also write the integral done by `compute_dPhi_dn` analytically and it comes out to this). However the numerical result converges to $0.03 + sin(\theta)$.

The only part of the numerical computation that is not performed analytically is the singular integral and the quadrature for the Fourier harmonics of $\Phi$ (what `jnp.linalg.solve` is basically doing under the hood). The latter is shown to converge to the correct result, so I can only conclude that the integration of the kernel is struggling. But there is no implementation bug in the kernel integration.

unalmis committed:
better comment in test

unalmis committed:
Remove image compare for now

unalmis committed:
Halfway through debugging

unalmis committed:
Add nfp loop wrapper

unalmis committed:
Fix laplace solver

unalmis committed:
Add markers for tests

unalmis committed:
Merge branch 'master' into ku/biot

unalmis committed:
Merge branch 'ku/biot' into dp/laplace

unalmis committed:
Finalize API for vacuum solver

unalmis committed:
Allow solving overdetermined system and passing remaining tests

unalmis committed:
Add warnings about JAX bug

unalmis commented on desc/integrals/_laplace.py:
```diff
@@ -0,0 +1,401 @@
+        Returns
+        -------
+        data : dict
+             Vacuum field ∇Φ stored in ``data["evl"]["grad(Phi)"]``.
```
if you prefer some other api let me know. i was taking care to make these functions pure (recall that is a technical term) so that they play nice with jax

unalmis replied:
ideally each method would be a registered compute function

unalmis commented on desc/integrals/singularities.py:
```diff
@@ -546,24 +543,13 @@ def eval_pt(eval_data_i):
-        "This is required to " + nfp_loop.__doc__,
-    )
-    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
-    f, _ = fori_loop(0, source_grid.NFP, nfp_loop, (f, source_data))
```
this seems to work without passing source_data in the f_data tuple. pointing out in case I'm missing something about pure functions

unalmis committed:
grad(Phi) not grad(phi)

unalmis committed:
Fix bug for setting default variable

unalmis committed:
Demonstrate #1599

dpanici committed:
Merge branch 'master' into ku/biot

dpanici commented on README.rst:
```diff
@@ -97,7 +97,7 @@ Contribute
 - `Issue Tracker <https://github.com/PlasmaControl/DESC/issues>`_
 - `Source Code <https://github.com/PlasmaControl/DESC/>`_
-- `Documentation <https://desc-docs.readthedocs.io/>`_
+- `Documentation <https://desc-docs.readthedocs.io/en/latest/>`_
```
I'd prefer this to be stable as then it is consistent with our pip package

unalmis replied:
ok

dpanici replied:
```suggestion
- `Documentation <https://desc-docs.readthedocs.io/en/stable/>`_
```

dpanici committed:
Merge branch 'master' into ku/biot

dpanici committed:
modify docstrings for fxns that dont use chunk_size

dpanici committed:
Merge branch 'ku/biot' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Rename laplace to vacuum

unalmis committed:
Use master compute data

unalmis committed:
Revert nfp loop

unalmis committed:
Remove thing moved to other pr

unalmis committed:
Use e^rho*sqrt(g) instead of n_rho

unalmis commented on desc/integrals/singularities.py:
```diff
@@ -805,12 +806,15 @@ def _kernel_1_over_r(eval_data, source_data, diag=False):
     dx = _dx(eval_data, source_data, diag)
-    n = rpz2xyz_vec(source_data["n_rho"], phi=source_data["phi"])
+    # Need to use e^rho*sqrt(g) to pass Green's ID test.
+    # Fourier spectrum is much more concentrated than n_rho for some reason.
```
otherwise the w7x one fails, which is interesting because there's many nodes and no truncation.

unalmis committed:
Update changelog

unalmis committed:
Fix mistake with sfaenormalize

unalmis committed:
Fix local_params func

unalmis committed:
same as previous commit

unalmis committed:
make sure int are used so that jax compiler works better

unalmis committed:
Use get_interpolatar func in free bdry

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Make sure errors and warnings print in expected color

unalmis committed:
Add mode_idx option ransform

unalmis committed:
Make inversion more efficient

unalmis committed:
Make inversion more efficient (#1637)

unalmis committed:
revert changes to transform

unalmis committed:
Add fixed point solve for Phi

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'dp/laplace' into ku/fp

unalmis committed:
Merge branch 'ku/cholesky' into dp/laplace

unalmis committed:
progressing

unalmis committed:
Add edited version of interp_2d from interpax

unalmis committed:
Add fixed point routine

unalmis committed:
Merge branch 'master' into dp/laplace

YigitElma committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
debugging fixed point

unalmis committed:
Better initial guess

unalmis committed:
Speed of matmul

unalmis committed:
j

dpanici commented on desc/basis.py:
```diff
@@ -657,57 +657,19 @@ def evaluate(self, grid, derivatives=np.array([0, 0, 0]), modes=None):
 
-        t = t[tidx]
-        z = z[zidx]
+        t = kwargs["t"] if "t" in kwargs else t[tidx]
```
`kwargs.get` would be cleaner here

unalmis replied:
then both branches would execute regardless

dpanici reviewed (dismissed):
I still should dig deeper but we need to decide on an API for this stuff. Currently it is its own class and creates/stores stuff in its own dictionaries. We don't have a good API for compute magnetic field stuff right now so there is not necessarily something to go off of, but I think if we can make it clear what this class takes and also how to access things (like, giving it a compute_magnetic_field function that would let it compute things off-surface and on-surface depending on a flag? or a compute_on_surface and compute_off_surface method) for computing B on and off-surface would be useful. I'd like it to at least have a `compute_magnetic_field` method which has similar API to our current stuff.

Potentially we could also make use of FourierCurrentPotentialField for the vacuum part of this field, since it is essentially similar to that representation, and that would allow re-use of those methods for the off-surface evaluations of B from the vacuum part.

unalmis commented:
> Potentially we could also make use of FourierCurrentPotentialField for the vacuum part of this field, since it is essentially similar to that representation, and that would allow re-use of those methods for the off-surface evaluations of B from the vacuum part.

FYI, the vacuum field cannot be generated from a surface current distribution on the boundary, i.e. a FourierCurrentPotentialField. The latter can only capture magnetic fields that are orthogonal to the boundary.

unalmis commented on desc/integrals/singularities.py:
```diff
@@ -765,11 +651,19 @@ def polar_pt(i):
 
 
+def _add_reduce(x):
+    # https://github.com/jax-ml/jax/issues/23493
```
only in new versions of jax it is resolved

unalmis commented on desc/integrals/singularities.py:
The primary reason for some of the changes here, it to ensure that only one large tensor product is computed when inverting the system for Phi. Otherwise you need to compute many large tensor products and that makes it slow. Recall JAX and JIT cannot optimize such matrix multiplication order.

unalmis replied:
And also so that I can compute integral of a non-periodic integrand.

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis commented on desc/integrals/singularities.py:
```diff
@@ -810,10 +712,16 @@ def singular_integral(
+        e.g. "Phi", and the second index should store the Python callable
+        that accepts a grid argument.
+    ndim : int
+        Default is kernel.ndim.
```
it is needed to make the integrals work for inversion without computing large tensor products

unalmis committed:
Fix grammer

unalmis commented on desc/compute/_surface.py:
```diff
@@ -21,14 +21,14 @@
 def _x_FourierRZToroidalSurface(params, transforms, profiles, data, **kwargs):
     R = transforms["R"].transform(params["R_lmn"])
     Z = transforms["Z"].transform(params["Z_lmn"])
     # TODO(#568): change when zeta no longer equals phi
```
most of the curve compute quantities don't work for general toroidal angle so i am just keeping them. this change was made to avoid some errors as the singular integral stuff assumed that if phi was computed then so was zeta and omega.

unalmis commented on desc/objectives/_free_boundary.py:
```diff
@@ -504,51 +504,20 @@ def build(self, use_jit=True, verbose=1):
-            interpolator = DFTInterpolator(
-                eval_grid, source_grid, self._st, self._sz, self._q
-            )
+        del self._st
```
they may not be the correct st and sz anymore. defensive programming

unalmis committed:
Second pass at new free boundary; taking break for the night

unalmis committed:
Fix import error

unalmis committed:
Simplified Bout computation in free boundary (#1600)

This PR goal is to separate the computation of the inner and outer
magnetic fields in the free boundary problem.

unalmis committed:
Add ability to evaluate at R phi Z in interior

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
off surface eval

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'ku/pos' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Change iterations to ones that should in theory converge

unalmis committed:
Add exterior solver

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
add missing key

unalmis committed:
Update alternative free bdry solver for left inverse surface gradient

unalmis committed:
speed up inversion

unalmis committed:
Bypass large tensor product

unalmis committed:
Handle secular terms of B_coil properly

unalmis committed:
Add axis limit

unalmis committed:
Update docstrings to warn about jax bug

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Pushing working commit

unalmis committed:
push updated test to sync with partner

dpanici commented on desc/integrals/_vacuum.py:
```diff
@@ -0,0 +1,693 @@
+            M=setdefault(Phi_M, Phi_grid.M),
+            N=setdefault(Phi_N, Phi_grid.N),
+            NFP=surface.NFP,
+            sym=setdefault(sym, False) and surface.sym,
```
This will fail if not False, as the options are False, cos, sin

dpanici committed:
fix sym bug

unalmis committed:
Update basis xyz

dpanici committed:
Merge branch 'dp/laplace' of github.com:PlasmaControl/DESC into dp/laplace

unalmis committed:
change resolution

unalmis committed:
Update README.rst

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
change exterior test so that singularity is outside domain and thus test converges and passes

dpanici committed:
Merge branch 'dp/laplace' of github.com:PlasmaControl/DESC into dp/laplace

unalmis committed:
Add test for splitting potential into secular and periodic parts

unalmis committed:
Debugging test

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
update

unalmis commented on tests/test_integrals.py:
```diff
@@ -692,28 +695,474 @@ def test_vanilla_params(self):
+        )
+
+    @pytest.mark.unit
+    def test_secular_splitting(self, chunk_size=100, I=0, Y=12):  # noqa: E741
```
@dpanici 

I am going to table this approach for now. When we want to add the effect of secular terms we can just compute the `B0` field that generates that secular term and use that method of splitting the potential as described in my report. That is the standard way of doing it and our functionality to that is already working and tested.

dpanici committed:
working commit for free bdry

unalmis commented on desc/integrals/_vacuum.py:
```diff
@@ -444,6 +445,38 @@ def compute_magnetic_field(self, chunk_size=None, coords=None):
+        self._data = self.compute_vacuum_field(chunk_size)
+        self._data["evl"]["B0+grad(Phi)"] = (
+            self._data["evl"]["B0"] + self._data["evl"]["grad(Phi)"]
+        )
```
This is the same as above function, maybe you copied accidentily?

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
no change

unalmis committed:
Updating naming for consistency with paper

unalmis committed:
Decreaes chunk size

unalmis committed:
fix failing test

dpanici commented on desc/integrals/_vacuum.py:
```diff
@@ -146,6 +146,7 @@ def __init__(
 
         # Compute data on source grid.
         self._src_transform = Transform(src_grid, basis, derivs=1)
+        self._evl_transform = Transform(evl_grid, basis, derivs=1)
```
This was for use inside of the free boundary so we can compute the data needed for the evl_data on the new eq surface at the current iteration in a way that works within our jitted compute functions (i..e to be passed to compute_fun inside of objective.compute)

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Updating sign conventions to match paper

unalmis committed:
fix comment

unalmis committed:
Remove singularity

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
non-singular Laplace solver and new magnetic field API (#1805)

unalmis committed:
Fix circular import

unalmis committed:
Fixing more imports

unalmis committed:
Remove out of date comment

unalmis committed:
Fix jit stuff from previous commit

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
resolving merge conflicts

unalmis committed:
Add options for convergence plots

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Add marker to skip plot test

unalmis committed:
Add another test

unalmis committed:
Update

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Remove old commetn

unalmis committed:
Fix failing test

unalmis committed:
Remove old warnings

unalmis committed:
jit fixed point method

unalmis committed:
Add plotting utils for convergence

unalmis committed:
fix comment

unalmis committed:
increase font size

unalmis committed:
Adds some compute quantities

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
update compute data for new quantities

unalmis committed:
use master compute data file

unalmis committed:
free surf parallel (#1822)

unalmis commented on desc/magnetic_fields/_laplace.py:
```diff
@@ -0,0 +1,398 @@
+        sym_coil=None,
+        B_coil=None,
+        Y_coil=None,
+        I_plasma=0.0,  # noqa: E741
```
```suggestion
        I_plasma=0.0,
```

unalmis commented on desc/objectives/_free_boundary.py:
```diff
@@ -884,6 +854,277 @@ def _print(fmt, fmax, fmin, fmean, f0max, f0min, f0mean, norm, unit):
+
+    _scalar = False
+    _print_value_fmt = "Free surface Error: "
+    _units = "T^2"
```
```suggestion
    _units = "T^2 m^2"
```

unalmis committed:
Apply suggestions from code review

unalmis committed:
Free surface potentials have stellarator symmetry

unalmis commented on desc/compute/_laplace.py:
```diff
@@ -0,0 +1,926 @@
+    data["B0*n"] = kwargs["B0"].compute_Bnormal(
+        kwargs["surface"],
+        eval_grid=grid,
+        source_grid=grid,
```
I am assuming all magnetic field objects copy the source grid to each field period and integrate over the full domain [0, 2pi).

unalmis commented on desc/compute/_laplace.py:
```diff
@@ -0,0 +1,926 @@
+    coords = jnp.column_stack([RpZ_data["R"], RpZ_data["phi"], RpZ_data["Z"]])
+    RpZ_data["B0"] = kwargs["B0"].compute_magnetic_field(
+        coords=coords,
+        source_grid=transforms["grid"],
```
I am assuming all magnetic field objects copy the source grid to each field period and integrate over the full domain [0, 2pi).

unalmis commented on desc/compute/_laplace.py:
```diff
@@ -0,0 +1,926 @@
+def _B_coil_field(params, transforms, profiles, data, **kwargs):
+    data["B_coil"] = kwargs["B_coil"].compute_magnetic_field(
+        coords=data["x"],
+        source_grid=transforms["grid"],
```
I am assuming all magnetic field objects copy the source grid to each field period and integrate over the full domain [0, 2pi).

unalmis committed:
Don't use eq.surface since that is not updated by optimizer

unalmis committed:
remove no longer needed code

unalmis committed:
Add coil grid as kwarg

unalmis committed:
Build psinv in get_transforms

dpanici committed:
Revert "working commit for free bdry"

This reverts commit 953d371d540be83f2bccc81c0d31ecb62c84abc5.

dpanici committed:
Merge branch 'dp/laplace' of github.com:PlasmaControl/DESC into dp/laplace

unalmis committed:
Add B_coil_chunk_size and don't interpolate gradients

dpanici committed:
Merge branch 'dp/laplace' of github.com:PlasmaControl/DESC into dp/laplace

dpanici committed:
Dp/laplace 2 (#1824)

unalmis committed:
Use 10000 instead of jnp.inf

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Add static attributes to freesurfaceerror objective

unalmis committed:
Remove old code

unalmis committed:
Reduce memory usage

unalmis committed:
Dummy commit to restart tests

unalmis committed:
Reduce chunk size in test

unalmis commented:
`nan` gradient is due to jax issue #14868

unalmis committed:
switch to lineax

unalmis committed:
Clean up if logic

unalmis committed:
Set well_posed to False for interiorNeumann because ill-conditioned

unalmis committed:
Add lineax tags

unalmis committed:
Don't assume integrals were computed with sufficient accuracy

unalmis committed:
same as previous commit

dpanici committed:
set well_posed=False to avoid NaNs

dpanici committed:
Merge branch 'dp/laplace' of github.com:PlasmaControl/DESC into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Add Neumann free surface

unalmis committed:
fix string for old python versions

unalmis committed:
Add warnings about symmetry

unalmis committed:
Fix spelling mistakes

unalmis committed:
Remove warnings about symmetry of basis

unalmis committed:
Update docstring in laplace.py

unalmis committed:
Update _laplace.py doc comment

unalmis committed:
Update free surface docstring

unalmis committed:
Update _free_boundary.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

unalmis committed:
Update _laplace.py kwarg docstring

unalmis committed:
Update _free_boundary.py docstring

unalmis committed:
Update comment in docstring

unalmis commented on desc/objectives/_free_boundary.py:
```diff
@@ -901,6 +871,369 @@ def _print(fmt, fmax, fmin, fmean, f0max, f0min, f0mean, norm, unit):
+            B_coil_chunk_size=self._B_coil_chunk_size,
+        )
+        # No net poloidal current in equation 4.13 of [1].
+        self._field.Y = 0.0 if self._is_neumann else data["Y_coil"]
```
As `obj.build` is not jitted, this line should execute and permanently modify self._field.Y.

unalmis commented on desc/objectives/_free_boundary.py:
```diff
@@ -901,6 +871,369 @@ def _print(fmt, fmax, fmin, fmean, f0max, f0min, f0mean, norm, unit):
+        maxiter=-1,
+        chunk_size=None,
+        B_coil_chunk_size=None,
+        I_sheet=0.0,
```
Reviewer should make this an optimizable parameter.

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Add some comments to test_exterior_Neumann

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Add unused dependency comment

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
merging

rahulgaur104 committed:
Merge branch 'master' into dp/laplace

dpanici reviewed (dismissed):
unicode characters cannot be easily typed, I would prefer changing all ∇φ to grad(Phi) etc so they can be written out easily

rahulgaur104 commented:
> All the solvers are tested for correctness. See the shared paper to explain this PR.
How can I access the shared paper? 
I have some simple questions about this process.

rahulgaur104 commented:
I am interested because there is an older paper that uses an almost similar idea (in terms of theory) to solve for a perturbed potential phi_1.
<img width="1414" height="2302" alt="Screenshot from 2025-08-22 13-58-11" src="https://github.com/user-attachments/assets/cfee70c1-9693-4d23-99b3-deab3eb5a71a" />

unalmis committed:
mark function as private

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Pulling changes downstream from https://github.com/f0uriest/interpax/pull/117

unalmis committed:
Increase coverage

unalmis committed:
Set to inf as that was not cause of nan

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
fix bad merge

unalmis committed:
increase codecov

unalmis committed:
Default to fixed point method and add stop tolerance option

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Cast theta, rho, zeta arrays to jax arrays

unalmis commented:
There aren't any automatically differentiable boundary integral equation solvers in Python. Hence, I would still like to merge this as the solver works fine.

unalmis committed:
Update singularities.py

unalmis committed:
Remove unused kernel and unused warn_dft kwarg

unalmis committed:
some plumbing to enable interpolation to less dense grid

unalmis committed:
fix master compue data

unalmis committed:
improve test

unalmis committed:
cosmetic change

unalmis committed:
same as previous commit

unalmis committed:
cast jax array to int

unalmis committed:
Add warning to objective to finish pull request.

unalmis committed:
Update changelog

and fix an unused kernel

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Trial 1: Merge branch 'master' into dp/laplace

unalmis committed:
Add missing test from master

unalmis committed:
Add cosmetic space deleted during merge

dpanici committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
Merge branch 'master' into dp/laplace

dpanici committed:
add anderson acceleration implementation

dpanici committed:
change default params

dpanici commented:
@dpanici  try using [jaxopt's anderson acceleration](https://jaxopt.github.io/stable/fixed_point.html#anderson-acceleration)

unalmis commented on desc/backend.py:
```diff
@@ -78,27 +79,98 @@ def g(x):
+            _, err, i, _, _ = state
+            return (i < m) & (~is_converged(err, tol))
+
+        ## run for m steps first with del2 then do anderson
```
### Caution with some of these changes

you must use iteration methods that guarentee convergence for this problem; no heuristics. Atkinson's `del2` method is a heuristic accelerator. It does NOT converge for our integral system can be seen by running `test_convergence_run_fixed_point`, and generating the plots with `test_convergence_plot_fixed_point`.

1. Standard fixed point contraction guarantees convergence.
2. Anderson method is guaranteed to converge, which is why I recommended it.
3. (If you know the eigenspectrum, a fully iterative Chebyshev method works, but eigenspectrum of this operator is open mathematics problem as I discussed in the shared paper. So don't use chebyshev iteration).

## Improvements that can be made
There are two main improvements that can be done in the fixed point method. The first is reducing iteration. We should see with Anderson because the function we iterate `f` calls singular integrals which are are so expensive, so doing some extra work with Anderson to reduce calls to `f` should be worthwile.

The second is that when we autodiff through this, the function `_tangent_solve` is computing the inverse on the right hand side of the implicit function theorem.
<img width="1027" height="176" alt="image" src="https://github.com/user-attachments/assets/f94d70ea-191c-4642-8d05-a9f11f96980a" />


That inversion is likely why optimization was slow/used memory with this implementation, even though computing the objective was fast. When we do normal root solve like with `map_pest_coordinates`, the Jacobian is 1x1 so its inverse is just the reciprocal. In this case the Jacobian is `mxm` where `m` is the number of output points the singular integral is evaluated ~`10^4`. We were using standard inversion so that's grows as m^3 so the cost was `10^6`. The way to avoid this is to also use fixed_point iterative to compute that inverse as discussed [here](https://docs.jax.dev/en/latest/advanced-autodiff.html#example-implicit-function-differentiation-of-iterative-implementations). Note this fixed point iteration to compute our vector Jacobian product doesn't involve any calls to singular integrals, so it should be super cheap. The jaxopt anderson method already has this built in with the methods `implicit_diff=True` and `implicit_diff_solve=None`. The latter refers to how that matrix is inverted. I did not check whether the default is fixed point for the latter. https://jaxopt.github.io/stable/_autosummary/jaxopt.AndersonAcceleration.html#jaxopt.AndersonAcceleration

unalmis replied:
> Could it be as simple as a change like this

not quite, but its the right idea.  I implemented it properly using lineax/optimistix in #2213 for big improvement in memory and compute.

dpanici committed:
update beta to not use relaxation, as convergence seems good without it

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
change default back to simple so tests pass, strange that with anderson the grad(phi) part of test fails despite Phi error going to machine precision

unalmis committed:
use code from interpax_fft

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
fix requirements.txt

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
merge

unalmis committed:
Merge branch 'master' into dp/laplace

dpanici committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Update requirements.txt

unalmis commented:
https://github.com/jax-ml/jax/pull/36383 was causing GPU results to be wrong on some machines. I have added a patch in an upstream library to bypass the issue.

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace

unalmis committed:
Merge branch 'master' into dp/laplace
